Lib/test/test_ntpath.py

```diff
@@ -637,6 +638,48 @@ def test_realpath_cwd(self):
         with os_helper.change_cwd(test_dir_short):
             self.assertPathEqual(test_file_long, ntpath.realpath("file.txt"))
 
+    @unittest.skipUnless(HAVE_GETFINALPATHNAME, 'need _getfinalpathname')
+    def test_realpath_permission(self):
+        # Test whether python can resolve the real filename of a
+        # shortened file name even if it does not have permission to access it.
+        ABSTFN = ntpath.realpath(os_helper.TESTFN)
+
+        os_helper.unlink(ABSTFN)
+        os_helper.rmtree(ABSTFN)
+        os.mkdir(ABSTFN)
+        self.addCleanup(os_helper.rmtree, ABSTFN)
+
+        test_file = ntpath.join(ABSTFN, "LongFileName123.txt")
+        test_file_short = ntpath.join(ABSTFN, "LONGFI~1.TXT")
+
+        with open(test_file, "wb") as f:
+            f.write(b"content")
+        # Automatic generation of short names may be disabled on
+        # NTFS volumes for the sake of performance.
+        # They're not supported at all on ReFS and exFAT.
+        DETACHED_PROCESS = 8
+        subprocess.run(
+            # Try to set the short name manually.
+            ['fsutil.exe', 'file', 'setShortName', test_file, 'LONGFI~1.TXT'],
+            creationflags=DETACHED_PROCESS
+        )
+
+        try:
+            self.assertPathEqual(test_file, ntpath.realpath(test_file_short))
+
+            # Deny the right to [S]YNCHRONIZE on the file to
+            # force GetFinalPathName to fail with ERROR_ACCESS_DENIED.
+            subprocess.run(
+                # bpo-30584: Use security identifier *S-1-5-32-545 instead
+                # of localized "Users" to not depend on the locale.
+                ['icacls.exe', test_file, '/deny', '*S-1-5-32-545:(S)'],
+                creationflags=DETACHED_PROCESS
+            )
+
+            self.assertPathEqual(test_file, ntpath.realpath(test_file_short))
+        except AssertionError:
+            raise unittest.SkipTest('the filesystem seems to lack support for short filenames')
```

Comment:
You can't consume the assertion from the second `assertPathEqual()` call. That's the actual test. Everything else is setup and checking whether the test is supported.

```suggestion
        try:
            self.assertPathEqual(test_file, ntpath.realpath(test_file_short))
        except AssertionError:
            raise unittest.SkipTest('the filesystem seems to lack support for short filenames')

        # Deny the right to [S]YNCHRONIZE on the file to all users (SDDL
        # alias "BU") to make opening it fail with ERROR_ACCESS_DENIED. 
        p = subprocess.run(['icacls.exe', test_file, '/deny', '*BU:(S)'],
                creationflags=subprocess.DETACHED_PROCESS)
        if p.returncode:
           raise unittest.SkipTest('failed to deny access to the test file') 

       self.assertPathEqual(test_file, ntpath.realpath(test_file_short))
```

Reply:
Right. Fixed.